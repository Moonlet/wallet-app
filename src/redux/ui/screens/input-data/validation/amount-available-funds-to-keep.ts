import BigNumber from 'bignumber.js';
import { Dispatch } from 'react';
import { IScreenFieldValidation } from '../../../../../components/widgets/types';
import { getBlockchain } from '../../../../../core/blockchain/blockchain-factory';
import { getChainId } from '../../../../preferences/selectors';
import { IReduxState } from '../../../../state';
import { getTokenConfig } from '../../../../tokens/static-selectors';
import { IAction } from '../../../../types';
import { getSelectedAccount } from '../../../../wallets/selectors';
import { setScreenInputValidation } from '../actions';

export const amountAvailableFundsToKeep = (
    validation: IScreenFieldValidation,
    field: string,
    flowId: string,
    getState: () => IReduxState,
    dispatch: Dispatch<IAction<any>>
) => {
    const state = getState();

    const account = getSelectedAccount(state);

    const inputData: any = state.ui.screens.inputData[flowId]?.data;

    const blockchain = account.blockchain;
    const blockchainInstance = getBlockchain(blockchain);
    const chainId = getChainId(state, blockchain);
    const token = account.tokens[chainId][blockchainInstance.config.coin];

    const tokenConfig = getTokenConfig(blockchain, token.symbol);

    // TODO

    const inputAmount = inputData?.amount;
    const inputAmountToStd = blockchainInstance.account.amountToStd(
        new BigNumber(inputAmount),
        tokenConfig.decimals
    );

    const screenAmount = inputData?.screenAmount;
    const screenAmountToStd = blockchainInstance.account.amountToStd(
        new BigNumber(screenAmount),
        tokenConfig.decimals
    );

    const minimumAmountToKeep = blockchainInstance.config.amountToKeepInAccount[account.type];

    if (
        screenAmountToStd.isLessThan(inputAmountToStd) &&
        inputAmountToStd.isLessThan(screenAmountToStd.plus(minimumAmountToKeep))
    ) {
        // Show error

        const fieldsErrors = [];

        for (const msgKey of Object.keys(validation?.messages || [])) {
            fieldsErrors.push(validation.messages[msgKey]);
        }

        setScreenInputValidation(flowId, {
            fieldsErrors: {
                [field]: fieldsErrors
            },
            valid: false
        })(dispatch, getState);
    }
};
