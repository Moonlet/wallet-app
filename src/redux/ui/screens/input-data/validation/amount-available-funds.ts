import BigNumber from 'bignumber.js';
import { Dispatch } from 'react';
import { IScreenFieldValidation } from '../../../../../components/widgets/types';
import { getBlockchain } from '../../../../../core/blockchain/blockchain-factory';
import { getChainId } from '../../../../preferences/selectors';
import { IReduxState } from '../../../../state';
import { getTokenConfig } from '../../../../tokens/static-selectors';
import { IAction } from '../../../../types';
import { getSelectedAccount } from '../../../../wallets/selectors';
import { setScreenInputValidation } from '../actions';

/**
 * This data is stored on screen key
 */
export const amountAvailableFunds = (
    validation: IScreenFieldValidation,
    field: string,
    screenKey: string,
    getState: () => IReduxState,
    dispatch: Dispatch<IAction<any>>
) => {
    const state = getState();

    const account = getSelectedAccount(state);

    const inputData: any = state.ui.screens.inputData[screenKey]?.data;

    const blockchain = account.blockchain;
    const blockchainInstance = getBlockchain(blockchain);
    const chainId = getChainId(state, blockchain);
    const token = account.tokens[chainId][blockchainInstance.config.coin];

    const tokenConfig = getTokenConfig(blockchain, token.symbol);

    let availableBalance = new BigNumber(token.balance?.available || '0');

    if (availableBalance.isGreaterThan(new BigNumber(0))) {
        availableBalance = availableBalance.minus(
            blockchainInstance.config.amountToKeepInAccount[account.type]
        );
    }

    const inputAmount = inputData?.amount;
    const inputAmountToStd = blockchainInstance.account.amountToStd(
        new BigNumber(inputAmount),
        tokenConfig.decimals
    );

    const regexMultipleDots = /(\..*){2,}/;

    if (inputAmountToStd.isGreaterThan(availableBalance) || regexMultipleDots.test(inputAmount)) {
        // Show error

        const fieldsErrors = [];

        for (const msgKey of Object.keys(validation?.messages || [])) {
            fieldsErrors.push(validation.messages[msgKey]);
        }

        setScreenInputValidation(screenKey, {
            fieldsErrors: {
                [field]: fieldsErrors
            },
            valid: false
        })(dispatch, getState);
    } else {
        // Valid input

        setScreenInputValidation(screenKey, {
            fieldsErrors: undefined,
            valid: true
        })(dispatch, getState);
    }
};

export const amountAvailableFundsToken = (
    validation: IScreenFieldValidation,
    field: string,
    screenKey: string,
    getState: () => IReduxState,
    dispatch: Dispatch<IAction<any>>
) => {
    const state = getState();

    const account = getSelectedAccount(state);

    const inputData = state.ui.screens.inputData[screenKey]?.data;

    const blockchain = account.blockchain;
    const blockchainInstance = getBlockchain(blockchain);

    const tokenSymbol = validation.params[0].tokenSymbol;
    const balance = new BigNumber(validation.params[0].balance);
    const swapType = validation.params[0]?.swapType;

    const inputAmount = inputData && inputData[field];

    const tokenConfig = getTokenConfig(blockchain, tokenSymbol);

    const screenValidations = state.ui.screens.inputData[screenKey]?.validation;
    const fieldsErrors = screenValidations?.fieldsErrors;
    let fields = (fieldsErrors && fieldsErrors[field]) || [];

    if (
        swapType &&
        state.ui.screens.inputData[screenKey]?.data?.swapType &&
        swapType !== state.ui.screens.inputData[screenKey]?.data?.swapType
    ) {
        // ignore this validations
        return;
    }

    const inputAmountToStd = blockchainInstance.account.amountToStd(
        new BigNumber(inputAmount),
        tokenConfig.decimals
    );

    if (inputAmountToStd.isGreaterThan(balance) || inputAmountToStd.isLessThanOrEqualTo(0)) {
        // Show error
        for (const msgKey of Object.keys(validation?.messages || [])) {
            const msg = validation.messages[msgKey] as any;
            // Make sure don't duplicate error messages
            let alreadyAdded = false;
            for (const f of fields || []) {
                if (JSON.stringify(f) === JSON.stringify(msg)) {
                    alreadyAdded = true;
                }
            }
            if (!alreadyAdded) fields.push(msg);
        }
    } else {
        // cleanup error messages
        fields = undefined;
    }

    let foundErrors = false;
    for (const fieldErrors of Object.values(fieldsErrors || {})) {
        if (fieldErrors && Array.isArray(fieldErrors) && fieldErrors.length !== 0) {
            foundErrors = true;
        }
    }

    if ((fields === undefined || fields?.length === 0) && foundErrors === false) {
        // All fields are valid => Validate Screen
        setScreenInputValidation(screenKey, {
            fieldsErrors: undefined,
            valid: true
        })(dispatch, getState);
    } else {
        // Set fields errors
        setScreenInputValidation(screenKey, {
            fieldsErrors: {
                ...fieldsErrors,
                [field]: fields
            },
            valid: false
        })(dispatch, getState);
    }
};
