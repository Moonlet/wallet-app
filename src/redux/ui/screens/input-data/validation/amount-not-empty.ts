import { Dispatch } from 'react';
import { IScreenFieldValidation } from '../../../../../components/widgets/types';
import { IReduxState } from '../../../../state';
import { IAction } from '../../../../types';
import { setScreenInputValidation } from '../actions';

export const inputAmountNotEmpty = (
    validation: IScreenFieldValidation,
    field: string,
    screenKey: string,
    getState: () => IReduxState,
    dispatch: Dispatch<IAction<any>>
) => {
    const state = getState();

    const inputData: any = state.ui.screens.inputData[screenKey]?.data;

    const amount = inputData?.amount || inputData?.input || (inputData && inputData[field]);

    if (
        !amount ||
        isNaN(amount) ||
        amount === '' ||
        amount === ' ' ||
        amount === undefined ||
        amount === null
    ) {
        // Show error

        const fieldsErrors = [];

        for (const msgKey of Object.keys(validation?.messages || [])) {
            fieldsErrors.push(validation.messages[msgKey]);
        }

        setScreenInputValidation(screenKey, {
            fieldsErrors: {
                [field]: fieldsErrors
            },
            valid: false
        })(dispatch, getState);
    }
};
