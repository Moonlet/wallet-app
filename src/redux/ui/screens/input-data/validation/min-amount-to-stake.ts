import BigNumber from 'bignumber.js';
import { Dispatch } from 'react';
import { IScreenFieldValidation } from '../../../../../components/widgets/types';
import { getBlockchain } from '../../../../../core/blockchain/blockchain-factory';
import { getChainId } from '../../../../preferences/selectors';
import { IReduxState } from '../../../../state';
import { getTokenConfig } from '../../../../tokens/static-selectors';
import { IAction } from '../../../../types';
import { getSelectedAccount } from '../../../../wallets/selectors';
import { setScreenInputValidation } from '../actions';

/**
 * This data is stored on screen key
 */
export const minAmountToStake = (
    validation: IScreenFieldValidation,
    field: string,
    screenKey: string,
    getState: () => IReduxState,
    dispatch: Dispatch<IAction<any>>
) => {
    const state = getState();

    const account = getSelectedAccount(state);

    const inputData: any = state.ui.screens.inputData[screenKey]?.data;

    const blockchain = account.blockchain;
    const blockchainInstance = getBlockchain(blockchain);
    const chainId = getChainId(state, blockchain);
    const token = account.tokens[chainId][blockchainInstance.config.coin];

    const tokenConfig = getTokenConfig(blockchain, token.symbol);

    const inputAmount = inputData?.amount;
    const inputAmountToStd = blockchainInstance.account.amountToStd(
        new BigNumber(inputAmount),
        tokenConfig.decimals
    );

    const minAmountDelegate = validation.params && validation.params[0]?.minAmountDelegate;

    if (inputAmountToStd.isLessThan(minAmountDelegate)) {
        // Show error

        const fieldsErrors = [];

        for (const msgKey of Object.keys(validation?.messages || [])) {
            fieldsErrors.push(validation.messages[msgKey]);
        }

        setScreenInputValidation(screenKey, {
            fieldsErrors: {
                [field]: fieldsErrors
            },
            valid: false
        })(dispatch, getState);
    }
};

export const minAmountToStakePerValidator = (
    validation: IScreenFieldValidation,
    field: string,
    screenKey: string,
    getState: () => IReduxState,
    dispatch: Dispatch<IAction<any>>
) => {
    const state = getState();

    const account = getSelectedAccount(state);

    const inputData: any = state.ui.screens.inputData[screenKey]?.data;

    const blockchain = account.blockchain;
    const blockchainInstance = getBlockchain(blockchain);
    const chainId = getChainId(state, blockchain);
    const token = account.tokens[chainId][blockchainInstance.config.coin];

    const tokenConfig = getTokenConfig(blockchain, token.symbol);

    const minAmountDelegate = validation.params && validation.params[0]?.minAmountDelegate;

    let allFieldsAreValid = true;

    for (const inputDataKey of Object.keys(inputData || {})) {
        if (
            inputData[inputDataKey] &&
            inputData[inputDataKey]?.amount &&
            inputData[inputDataKey]?.validator
        ) {
            const amount = inputData[inputDataKey].amount;
            const amountToStd = blockchainInstance.account.amountToStd(
                new BigNumber(amount),
                tokenConfig.decimals
            );

            if (amountToStd.isLessThan(minAmountDelegate) && amountToStd.isGreaterThan(0)) {
                // Show error

                const fieldsErrors = [];

                for (const msgKey of Object.keys(validation?.messages || [])) {
                    fieldsErrors.push(validation.messages[msgKey]);
                }

                setScreenInputValidation(screenKey, {
                    fieldsErrors: {
                        [field]: fieldsErrors
                    },
                    valid: false
                })(dispatch, getState);

                allFieldsAreValid = false;
            }
        }
    }

    if (allFieldsAreValid) {
        // Valid input

        setScreenInputValidation(screenKey, {
            fieldsErrors: undefined,
            valid: true
        })(dispatch, getState);
    }
};
