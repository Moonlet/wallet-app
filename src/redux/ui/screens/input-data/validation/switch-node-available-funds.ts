import BigNumber from 'bignumber.js';
import { Dispatch } from 'react';
import { IScreenFieldValidation } from '../../../../../components/widgets/types';
import { getBlockchain } from '../../../../../core/blockchain/blockchain-factory';
import { getChainId } from '../../../../preferences/selectors';
import { IReduxState } from '../../../../state';
import { getTokenConfig } from '../../../../tokens/static-selectors';
import { IAction } from '../../../../types';
import { getSelectedAccount } from '../../../../wallets/selectors';
import { setScreenInputValidation } from '../actions';

export const switchNodeAvailableFunds = (
    validation: IScreenFieldValidation,
    field: string,
    screenKey: string,
    getState: () => IReduxState,
    dispatch: Dispatch<IAction<any>>
) => {
    const state = getState();

    const account = getSelectedAccount(state);

    const inputData: any = state.ui.screens.inputData[screenKey]?.data;

    const blockchain = account.blockchain;
    const blockchainInstance = getBlockchain(blockchain);
    const chainId = getChainId(state, blockchain);
    const token = account.tokens[chainId][blockchainInstance.config.coin];

    const tokenConfig = getTokenConfig(blockchain, token.symbol);

    const flowId = validation.params[0].flowId;

    const nodeAvailableAmount = new BigNumber(
        (state.ui.screens.inputData &&
            state.ui.screens.inputData[flowId]?.data?.switchNodeValidator?.availableBalance) ||
            '0'
    );

    const inputAmount = blockchainInstance.account.amountToStd(
        new BigNumber(inputData?.amount),
        tokenConfig.decimals
    );

    if (inputAmount.isGreaterThan(nodeAvailableAmount)) {
        // Show error

        const fieldsErrors = [];

        for (const msgKey of Object.keys(validation?.messages || [])) {
            fieldsErrors.push(validation.messages[msgKey]);
        }

        setScreenInputValidation(screenKey, {
            fieldsErrors: {
                [field]: fieldsErrors
            },
            valid: false
        })(dispatch, getState);
    } else {
        // Valid input

        setScreenInputValidation(screenKey, {
            fieldsErrors: undefined,
            valid: true
        })(dispatch, getState);
    }
};
