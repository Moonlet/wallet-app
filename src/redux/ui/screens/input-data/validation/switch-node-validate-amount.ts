import BigNumber from 'bignumber.js';
import { cloneDeep } from 'lodash';
import { Dispatch } from 'react';
import { IScreenFieldValidation } from '../../../../../components/widgets/types';
import { getBlockchain } from '../../../../../core/blockchain/blockchain-factory';
import { formatNumber } from '../../../../../core/utils/format-number';
import { getChainId } from '../../../../preferences/selectors';
import { IReduxState } from '../../../../state';
import { getTokenConfig } from '../../../../tokens/static-selectors';
import { IAction } from '../../../../types';
import { getSelectedAccount } from '../../../../wallets/selectors';
import { setScreenInputValidation } from '../actions';

/**
 * This data is stored on screen key
 */
export const switchNodeValidateAmount = (
    validation: IScreenFieldValidation,
    field: string,
    screenKey: string,
    getState: () => IReduxState,
    dispatch: Dispatch<IAction<any>>
) => {
    const state = getState();

    const account = getSelectedAccount(state);

    const inputData: any = state.ui.screens.inputData[screenKey]?.data;

    const blockchain = account.blockchain;
    const blockchainInstance = getBlockchain(blockchain);
    const chainId = getChainId(state, blockchain);
    const token = account.tokens[chainId][blockchainInstance.config.coin];

    const tokenConfig = getTokenConfig(blockchain, token.symbol);

    const inputAmount = blockchainInstance.account.amountToStd(
        new BigNumber(inputData?.amount),
        tokenConfig.decimals
    );

    const flowId = validation.params[0].flowId;
    const minAmountDelegate = new BigNumber(validation.params[0].minAmountDelegate);

    const nodeAvailableAmount = new BigNumber(
        (state.ui.screens.inputData &&
            state.ui.screens.inputData[flowId]?.data?.switchNodeValidator?.availableBalance) ||
            '0'
    );

    if (
        nodeAvailableAmount.minus(minAmountDelegate).isLessThan(inputAmount) &&
        inputAmount.isLessThan(nodeAvailableAmount)
    ) {
        // Show error

        const fieldsErrors = [];

        for (const msgKey of Object.keys(validation?.messages || [])) {
            const msg = cloneDeep(validation.messages[msgKey]);
            msg.message =
                msg.message +
                ` ${formatNumber(
                    blockchainInstance.account.amountFromStd(
                        nodeAvailableAmount.minus(minAmountDelegate),
                        tokenConfig.decimals
                    ),
                    { currency: blockchainInstance.config.coin }
                )}`;
            fieldsErrors.push(msg);
        }

        setScreenInputValidation(screenKey, {
            fieldsErrors: {
                [field]: fieldsErrors
            },
            valid: false
        })(dispatch, getState);
    }
};
