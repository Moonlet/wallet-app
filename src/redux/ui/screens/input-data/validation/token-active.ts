import { Dispatch } from 'react';
import { Dialog } from '../../../../../components/dialog/dialog';
import { IScreenFieldValidation } from '../../../../../components/widgets/types';
import { translate } from '../../../../../core/i18n';
import { NavigationService } from '../../../../../navigation/navigation-service';
import { getChainId } from '../../../../preferences/selectors';
import { IReduxState } from '../../../../state';
import { IAction } from '../../../../types';
import { getSelectedAccount } from '../../../../wallets/selectors';
import { setScreenInputValidation } from '../actions';

export const tokenActiveWallet = (
    validation: IScreenFieldValidation,
    field: string,
    screenKey: string,
    getState: () => IReduxState,
    dispatch: Dispatch<IAction<any>>
) => {
    const state = getState();

    const account = getSelectedAccount(state);
    const chainId = getChainId(state, account.blockchain);

    const symbol = validation?.params && validation.params[0]?.symbol;

    const tokens = account.tokens[chainId];

    if (symbol && tokens) {
        if (!tokens[symbol]) {
            // TODO
        }

        if (tokens[symbol]) {
            const tokenConfig = tokens[symbol];
            if (!tokenConfig.active) {
                Dialog.alert(
                    'Inactive token',
                    `Please activate ${symbol} token`,
                    {
                        text: translate('App.labels.cancel'),
                        onPress: () => {
                            //
                        }
                    },
                    {
                        text: 'Activate',
                        onPress: () => NavigationService.navigate('AddToken', {})
                    }
                );

                // Show error

                const fieldsErrors = [];

                for (const msgKey of Object.keys(validation?.messages || [])) {
                    fieldsErrors.push(validation.messages[msgKey]);
                }

                setScreenInputValidation(screenKey, {
                    fieldsErrors: {
                        [field]: fieldsErrors
                    },
                    valid: false
                })(dispatch, getState);
            }
        }
    }
};
