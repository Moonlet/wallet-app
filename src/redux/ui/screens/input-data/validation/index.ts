import { amountAvailableFunds, amountAvailableFundsToken } from './amount-available-funds';
import { amountAvailableFundsToKeep } from './amount-available-funds-to-keep';
import { minAmountToStake, minAmountToStakePerValidator } from './min-amount-to-stake';
import { switchNodeMinAmountDelegate } from './switch-node-min-amount-delegate';
import { switchNodeValidateAmount } from './switch-node-validate-amount';
import { switchNodeAvailableFunds } from './switch-node-available-funds';
import { inputAmountNotEmpty } from './amount-not-empty';
import { tokenActiveWallet } from './token-active';

export const screenInputValidationActions = {
    amountAvailableFunds,
    amountAvailableFundsToken,
    amountAvailableFundsToKeep,
    minAmountToStake,
    minAmountToStakePerValidator,
    switchNodeMinAmountDelegate,
    switchNodeValidateAmount,
    switchNodeAvailableFunds,
    inputAmountNotEmpty,
    tokenActiveWallet
};
