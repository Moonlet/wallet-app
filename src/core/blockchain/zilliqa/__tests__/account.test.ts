import { Zilliqa } from '../';
import {
    getPubKeyFromPrivateKey,
    getAddressFromPrivateKey,
    getAddressFromPublicKey
} from '@zilliqa-js/crypto/dist/util'; // import like this to optimize imports
import { toBech32Address } from '@zilliqa-js/crypto/dist/bech32';
import { isBech32 } from '@zilliqa-js/util/dist/validation';
import BigNumber from 'bignumber.js';

jest.mock('@zilliqa-js/crypto/dist/util', () => {
    const util = jest.requireActual('@zilliqa-js/crypto/dist/util');
    return {
        ...util,
        getPubKeyFromPrivateKey: jest.fn().mockReturnValue('PUBLIC_KEY'),
        getAddressFromPrivateKey: jest.fn().mockReturnValue('address'),
        getAddressFromPublicKey: jest.fn().mockReturnValue('address'),
        amountToStd: jest.fn().mockReturnValue(new BigNumber(0)),
        amountFromStd: jest.fn().mockReturnValue(new BigNumber(0))
    };
});

jest.mock('@zilliqa-js/util/dist/validation', () => ({
    isBech32: jest.fn().mockReturnValue(true)
}));

jest.mock('@zilliqa-js/crypto/dist/bech32', () => ({
    toBech32Address: jest.fn().mockReturnValue('BECH32_ADDRESS')
}));

const clearMocks = () => {
    // @ts-ignore
    getPubKeyFromPrivateKey.mockClear();
    // @ts-ignore
    getAddressFromPrivateKey.mockClear();
    // @ts-ignore
    getAddressFromPublicKey.mockClear();
    // @ts-ignore
    toBech32Address.mockClear();
    // @ts-ignore
    isBech32.mockClear();
    // @ts-ignore
    amountToStd.mockClear();
    // @ts-ignore
    amountFromStd.mockClear();
};

describe('Zilliqa account', () => {
    beforeEach(() => {
        clearMocks();
    });

    test('isValidChecksumAddress()', () => {
        expect(Zilliqa.account.isValidChecksumAddress('ADDRESS')).toBe(true);
        expect(isBech32).toBeCalledWith('ADDRESS');
    });

    test('isValidAddress()', () => {
        expect(Zilliqa.account.isValidAddress('ADDRESS')).toBe(true);
        expect(isBech32).toBeCalledWith('ADDRESS');
    });

    test('publicToAddress()', () => {
        expect(Zilliqa.account.publicToAddress('PUBLIC_KEY')).toBe('BECH32_ADDRESS');
        expect(getAddressFromPublicKey).toBeCalledWith('PUBLIC_KEY');
        expect(toBech32Address).toBeCalledWith('address');
    });

    test('privateToPublic()', () => {
        expect(Zilliqa.account.privateToPublic('PRIVATE_KEY')).toBe('PUBLIC_KEY');
        expect(getPubKeyFromPrivateKey).toBeCalledWith('PRIVATE_KEY');
    });

    test('privateToAddress()', () => {
        expect(Zilliqa.account.privateToAddress('PRIVATE_KEY')).toBe('BECH32_ADDRESS');
        expect(getAddressFromPrivateKey).toBeCalledWith('PRIVATE_KEY');
        expect(toBech32Address).toBeCalledWith('address');
    });

    test('getAccountFromPrivateKey()', () => {
        // valid private key
        const result = Zilliqa.account.getAccountFromPrivateKey('PRIVATE_KEY', 1);
        expect(getPubKeyFromPrivateKey).toBeCalledWith('PRIVATE_KEY');
        expect(getAddressFromPrivateKey).toBeCalledWith('PRIVATE_KEY');
        expect(toBech32Address).toBeCalledWith('address');

        expect(result).toMatchSnapshot();

        // TODO: test with invalid privatekey, without mocks
    });
});
