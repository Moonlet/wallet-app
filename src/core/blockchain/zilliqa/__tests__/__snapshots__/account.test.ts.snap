// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Zilliqa account getAccountFromPrivateKey() 1`] = `
Object {
  "address": "BECH32_ADDRESS",
  "blockchain": "ZILLIQA",
  "index": 1,
  "publicKey": "PUBLIC_KEY",
  "selected": false,
  "tokens": Object {
    "ZIL": Object {
      "active": true,
      "decimals": 12,
      "logo": Object {
        "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/zil.png",
      },
      "name": "Zilliqa",
      "order": 0,
      "symbol": "ZIL",
      "type": "NATIVE",
      "ui": Object {
        "decimals": 3,
        "tokenScreenComponent": "DEFAULT",
      },
      "units": Object {
        "LI": "1000000",
        "QA": "1",
        "ZIL": "1000000000000",
      },
    },
  },
}
`;
